<a href="https://colab.research.google.com/github/PauSerranoHerraiz/S7-01-Pau-Serrano/blob/main/S8_01_Pau_Serrano.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color="green">  Nivell 1 - EDA estructurat i neteja bàsica </font>

## <font color = "pink"> Exercici 1: Exploració inicial del dataset <font>
🎯 Objectiu: entendre l’estructura general del dataset i detectar possibles problemes abans de començar a netejar o transformar dades.

En aquest exercici has de realitzar una exploració inicial del DataFrame per respondre, com a mínim, a les següents preguntes:


» Quantes files i columnes té el dataset?

» Quin tipus de dades conté cada columna?

» Hi ha files duplicades?

» Les dades semblen coherents a primera vista?

» Detectes possibles problemes o inconsistències evidents?

Per fer-ho, hauràs de:


» Analitzar les dimensions del dataset

» Mostrar una mostra representativa de files

» Revisar els tipus de variables

» Les dades semblen coherents a primera vista?

» Detectar files duplicades

» Identificar i descriure possibles problemes inicials (sense solucionar-los encara)

 ### Important

acompanya el codi amb un breu comentari explicant què has observat i quins problemes creus que caldrà abordar en els exercicis següents.

In [5]:
import pandas as pd
import unicodedata
from google.colab import files

# Pujo l'arxiu
uploaded = files.upload()
empleats = list(uploaded.keys())[0]

# Carreguem el dataset i ens saltem les 3 primeres files, indicant que el separador és el punt i coma.
df = pd.read_csv(
    empleats,
    skiprows=3,
    sep=",",
    header=0
)

def normalitzar_columnes(df):

    noves = []

    for c in df.columns:

        c = str(c).strip().lower()

        c = unicodedata.normalize(
            "NFKD",
            c
        ).encode(
            "ascii",
            "ignore"
        ).decode()

        c = c.replace(" ", "_")

        noves.append(c)

    df.columns = noves

    return df

df = normalitzar_columnes(df)

print(df.columns)



print(df["salari_mensual"].dtype)

print(df["salari_mensual"].describe())

# Dimensions del dataset
print("Dimensions del dataset:")
print(df.shape)

# Files i columnes
print(f"\nNombre de files: {df.shape[0]}")
print(f"Nombre de columnes: {df.shape[1]}")

# Primeres files
print("\nPrimeres 5 files:")
display(df.head())

# Últimes files
print("\nÚltimes 5 files:")
display(df.tail())

# Informació general
print("\nInformació del dataset:")
df.info()

# Tipus de dades
print("\nTipus de dades:")
display(df.dtypes)

# Estadístiques descriptives
print("\nEstadístiques de variables numèriques:")
display(df.describe())

# Estadístiques variables categòriques
print("\nVariables categòriques:")
display(df.describe(include="object"))

# Valors nuls
print("\nValors nuls per columna:")
display(df.isnull().sum())

# Files duplicades
duplicats = df.duplicated().sum()

print(f"\nNombre de files duplicades: {duplicats}")


Saving sprint10_complex.csv to sprint10_complex (2).csv
Index(['unnamed:_0', 'nom', 'cognoms', 'dni', 'pais_d'origen', 'ciutat',
       'dia_de_naixement', 'mes_de_naixement', 'any_de_naixement', 'genere',
       'salari_mensual', 'fills', 'no_fills', 'grup_professional', 'ca_rrec',
       'nombre_fills', 'te_cotxe', 'km_anuals', 'consum_mitja_l_100km',
       'temperatura_mitjana_ciutat'],
      dtype='object')
object
count     1006
unique     855
top        n/d
freq        14
Name: salari_mensual, dtype: object
Dimensions del dataset:
(1007, 20)

Nombre de files: 1007
Nombre de columnes: 20

Primeres 5 files:


,unnamed:_0,nom,cognoms,dni,pais_d'origen,ciutat,dia_de_naixement,mes_de_naixement,any_de_naixement,genere,salari_mensual,fills,no_fills,grup_professional,ca_rrec,nombre_fills,te_cotxe,km_anuals,consum_mitja_l_100km,temperatura_mitjana_ciutat
0,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1



Últimes 5 files:


,unnamed:_0,nom,cognoms,dni,pais_d'origen,ciutat,dia_de_naixement,mes_de_naixement,any_de_naixement,genere,salari_mensual,fills,no_fills,grup_professional,ca_rrec,nombre_fills,te_cotxe,km_anuals,consum_mitja_l_100km,temperatura_mitjana_ciutat
1002,1002,Alba,Muñoz,97954019T,Espanya,Bilbao,25.0,10.0,1983.0,D,"1,409 €",NaN,True,Grup B,Responsable de vendes,0.0,NaN,NaN,NaN,14.7
1003,1003,Claudia,Gil Ferrer,55297384F,Espanya,València,1.0,11.0,1981.0,H,1.789 €,NaN,True,Grup B,Responsable de vendes,0.0,True,29952.0,9.2,20.3
1004,1004,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
1005,1005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1006,1006,???,###,XXXX,Narnia,Desconeguda,-99.0,99.0,3000.0,??,molts,NaN,NaN,Z,???,NaN,NaN,NaN,NaN,NaN



Informació del dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   unnamed:_0                  1007 non-null   int64  
 1   nom                         1006 non-null   object 
 2   cognoms                     1006 non-null   object 
 3   dni                         1006 non-null   object 
 4   pais_d'origen               1006 non-null   object 
 5   ciutat                      1006 non-null   object 
 6   dia_de_naixement            1006 non-null   float64
 7   mes_de_naixement            1006 non-null   float64
 8   any_de_naixement            1006 non-null   float64
 9   genere                      966 non-null    object 
 10  salari_mensual              1006 non-null   object 
 11  fills                       419 non-null    object 
 12  no_fills                    585 non-null    object 
 13  grup_pro

,0
unnamed:_0,int64
nom,object
cognoms,object
dni,object
pais_d'origen,object
ciutat,object
dia_de_naixement,float64
mes_de_naixement,float64
any_de_naixement,float64
genere,object



Estadístiques de variables numèriques:


,unnamed:_0,dia_de_naixement,mes_de_naixement,any_de_naixement,nombre_fills,km_anuals,consum_mitja_l_100km,temperatura_mitjana_ciutat
count,1007.000000,1006.000000,1006.000000,1006.000000,709.000000,315.000000,261.000000,936.000000
mean,503.000000,15.027833,8.556660,2002.229622,1.203103,55951.050794,11.024521,15.379167
std,290.840162,10.327436,14.298532,143.118077,1.389404,62095.437005,8.775557,7.276425
min,0.000000,-99.000000,0.000000,1940.000000,0.000000,-50.000000,0.500000,-10.000000
25%,251.500000,7.000000,3.250000,1969.000000,0.000000,22632.000000,4.700000,12.800000
50%,503.000000,15.000000,6.000000,1982.000000,1.000000,56135.000000,9.800000,15.500000
75%,754.500000,24.000000,10.000000,1995.000000,2.000000,80000.000000,13.600000,18.300000
max,1006.000000,32.000000,99.000000,3000.000000,5.000000,1000000.000000,25.000000,42.000000



Variables categòriques:


,nom,cognoms,dni,pais_d'origen,ciutat,genere,salari_mensual,fills,no_fills,grup_professional,ca_rrec,te_cotxe
count,1006,1006,1006,1006,1006,966,1006,419,585,1005,1005,658
unique,103,468,991,6,45,9,855,1,2,5,14,2
top,AdriÃ,Ferrer,90585882A,Espanya,Palma,D,n/d,True,True,Grup A,Data Analyst,False
freq,34,22,6,719,58,325,14,419,584,466,142,330



Valors nuls per columna:


,0
unnamed:_0,0
nom,1
cognoms,1
dni,1
pais_d'origen,1
ciutat,1
dia_de_naixement,1
mes_de_naixement,1
any_de_naixement,1
genere,41



Nombre de files duplicades: 0


### <font color = "yellow"> Problmes que caldrà abordar:

* Columna Fantasma: Hi ha una primera columna (Unnamed: 0) que és simplement l'índex anterior exportat per error i no aporta valor analític.

* Problemes de codificació: Caràcters com accents i "ñ" s'estan llegint malament (ex: Cà rrec, Muñoz).

* "Salari mensual" com a text (String): Conté caràcters de text barrejats amb números ("1,469 €", "1358 euros", comes de milers). Serà impossible fer mitjanes ni sumar aquesta columna si no es converteix exclusivament a número.

* Inconsistència de categories: A Gènere veiem que es barreja "Dona", "D", i "H". S'ha d'estadaritzar.

* Redundància en la variable de "Fills": Hi ha tres columnes diferents (Fills, No Fills, i Nombre_fills) que guarden informació creuada sobre el mateix concepte, incloent-hi molts valors nuls (NaN) i valors booleans (True). S'ha de fer neteja.

* Molts valors nuls (NaN) dispersos: Hi ha dades buides evidents en variables com Km_anuals o Te_cotxe, que s'han de tractar.

-----------------------

## <font color = "pink"> Exercici 2: Validacions bàsiques de les dades i decisions d'eliminació </font>
🎯 Objectiu: assegurar que les dades compleixen unes condicions mínimes de qualitat i determinar quines files poden continuar sent analitzades i quines s'han de descartar.

En aquest exercici hauràs de validar aspectes clau del dataset que són crítics en qualsevol context real d'anàlisi de dades, prenent decisions justificades sobre la validesa o eliminació de registres.


Aspectes a validar:


» Existeixen identificadors únics (DNI) per a cada persona? <font color = "yellow"> Sí, tots els registres tenen el seu DNI </font>

» Què faràs amb les files duplicades o registres repetits? <font color = "yellow"> Si estàn exatament duplicades en tots els seus camps, eliminar-ne una. Si està duplicada la PK(DNI) però la resta no concorda, eliminar tots els registres amnb la mateixa PK, ja que es impossible discernir quin d'ells es el correcte </font>

» Hi ha valors obligatoris que falten o informació incoherent? <font color ="yellow">Si. Hi ha camps buits </font>

» Detectes formats incorrectes, errors de codificació o valors impossibles?

Concretament, hauràs de:

» Corregir possibles problemes de codificació i accentuació en els noms de les columnes i en els valors de text, si escau.

» Detectar i eliminar files duplicades exactes, indicant quants registres s'han eliminat.

» Comprovar la unicitat del DNI. En cas de detectar DNIs duplicats, hauràs d'analitzar la coherència de la informació associada i descartar o anul·lar els registres que no siguin fiables, justificant clarament la decisió presa.

» Identificar els camps clau que no haurien de contenir valors nuls, detectar registres amb informació insuficient i eliminar únicament aquelles files completament buides o amb informació irrecuperable o no substituïble, explicant per què no és possible ni raonable imputar-les.

 Important


en aquest exercici no s'espera que imputis dades, sinó que decideixis quins registres poden formar part de l'anàlisi i quins no, argumentant sempre el criteri utilitzat.


🧠 Si el text "s'entén però està mal escrit" → codificació


Si el text "està ben escrit però no és útil per a l'anàlisi" → validació

In [6]:

import pandas as pd


# 1. ELIMINAR DUPLICATS EXACTES

files_abans = len(df)

df = df.drop_duplicates()

files_despres = len(df)

print(f"Duplicats exactes eliminats: {files_abans - files_despres}")

# 2. COMPROVAR DNI DUPLICATS

dni_duplicats = df[df.duplicated(subset="dni", keep=False)]

print(f"\nNombre de registres amb DNI duplicat: {len(dni_duplicats)}")

if len(dni_duplicats) > 0:

    display(
        dni_duplicats.sort_values("dni")
    )

    # Comptem quantes dades té cada registre
    df["completitud"] = df.notna().sum(axis=1)

    # Conservem el registre amb més informació
    df = (
        df
        .sort_values(
            by=["dni", "completitud"],
            ascending=[True, False]
        )
        .drop_duplicates(
            subset="dni",
            keep="first"
        )
    )

    # Eliminem la columna auxiliar
    df.drop(columns="completitud", inplace=True)

print(f"\nNombre de registres després de revisar els DNI: {len(df)}")

# 3. DETECTAR VALORS NULS

print("\nValors nuls per columna:")

display(
    df.isnull()
      .sum()
      .sort_values(ascending=False)
)

# 4. ELIMINAR REGISTRES IRRECUPERABLES

camps_obligatoris = [
    "dni",
    "nom",
    "cognoms",
    "genere"
]

registres_invalids = df[
    df[camps_obligatoris]
      .isnull()
      .any(axis=1)
]

print("\nRegistres eliminats per manca d'informació obligatòria:")

display(registres_invalids)

df = df.dropna(
    subset=camps_obligatoris,
    how="any"
)

print(f"\nNombre de registres després d'eliminar els registres invàlids: {len(df)}")

# 5. FILES COMPLETAMENT BUIDES


files_buides = df.isnull().all(axis=1).sum()

print(f"\nFiles completament buides: {files_buides}")

# 6. RESUM FINAL

print("=" * 50)
print("RESUM DE LA VALIDACIÓ")
print("=" * 50)

print(f"Nombre final de registres: {len(df)}")
print(f"DNI únics: {df['dni'].nunique()}")

print("\nValors nuls restants:")

display(df.isnull().sum())

Duplicats exactes eliminats: 0

Nombre de registres amb DNI duplicat: 24


,unnamed:_0,nom,cognoms,dni,pais_d'origen,ciutat,dia_de_naixement,mes_de_naixement,any_de_naixement,genere,salari_mensual,fills,no_fills,grup_professional,ca_rrec,nombre_fills,te_cotxe,km_anuals,consum_mitja_l_100km,temperatura_mitjana_ciutat
604,604,Laia,Navarro,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,"1,676 €",True,False,Grup A,Analista,1.0,True,50.0,NaN,14.2
250,250,Laia,Navarro Vila,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,1235 euros,True,True,Grup D,Analista junior,0.0,False,NaN,NaN,10.3
759,759,Oriol,Gómez,16618423V,Espanya,Sevilla,10.0,1.0,1968.0,NaN,"2,000 €",NaN,NaN,Grup A,Responsable de vendes,5.0,NaN,NaN,NaN,20.7
758,758,Oriol,Gómez,16618423V,Espanya,Mà laga,10.0,1.0,1968.0,H,1879 euros,True,NaN,Grup B,Data Analyst,4.0,False,NaN,NaN,NaN
754,754,Oriol,Gómez,16618423V,Espanya,Barcelona,10.0,1.0,1968.0,Home,2.358 €,True,NaN,NaN,NaN,4.0,NaN,NaN,NaN,NaN
654,654,Oriol,Gómez,16618423V,Espanya,Madrid,10.0,1.0,1968.0,Home,1731 euros,True,NaN,Grup B,Data analyst,4.0,NaN,NaN,NaN,NaN
692,692,Alexia,Pérez Díaz,49529438G,Espanya,Mà laga,12.0,10.0,1981.0,Dona,1663€,NaN,True,Grup C,Cap de projecte,NaN,True,80000.0,10.7,16.9
558,558,Alexia,Pérez Díaz,49529438G,Espanya,Mà laga,12.0,10.0,1981.0,Dona,"1,176 €",NaN,True,Grup A,Data Analyst,0.0,False,NaN,NaN,38.0
757,757,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
1004,1004,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9



Nombre de registres després de revisar els DNI: 992

Valors nuls per columna:


,0
consum_mitja_l_100km,733
km_anuals,679
fills,578
no_fills,417
te_cotxe,344
nombre_fills,297
temperatura_mitjana_ciutat,69
genere,40
salari_mensual,1
dni,1



Registres eliminats per manca d'informació obligatòria:


,unnamed:_0,nom,cognoms,dni,pais_d'origen,ciutat,dia_de_naixement,mes_de_naixement,any_de_naixement,genere,salari_mensual,fills,no_fills,grup_professional,ca_rrec,nombre_fills,te_cotxe,km_anuals,consum_mitja_l_100km,temperatura_mitjana_ciutat
126,126,Lucas,Petit,17324144F,França,Paris,21.0,5.0,1983.0,NaN,2227€,True,NaN,Grup C,Responsable de vendes,1.0,False,NaN,NaN,12.6
102,102,Noa,Ruiz Pérez,18113477L,Espanya,Murcia,12.0,8.0,1990.0,NaN,"2,627 €",True,NaN,Grup D,Responsable de vendes,1.0,False,NaN,NaN,22.2
523,523,Marc,Ferrer,20402470N,Espanya,Las Palmas de Gran Canaria,23.0,7.0,1964.0,NaN,929€,True,NaN,Grup A,Data Analyst,4.0,True,50000.0,0.5,20.3
325,325,Marina,Álvarez Puig,20880123Z,Espanya,Valladolid,0.0,0.0,2019.0,NaN,"1,320 €",True,NaN,Grup A,Responsable de vendes,1.0,False,NaN,NaN,6.2
543,543,AdriÃ,Gómez Soler,21091816B,Espanya,Bilbao,26.0,8.0,1964.0,NaN,830 €,True,NaN,Grup A,Cap de Projecte,1.0,False,NaN,NaN,10.9
61,61,Xavier,Muñoz Roca,21370793D,Espanya,València,20.0,3.0,1983.0,NaN,3016€,NaN,True,Grup D,Tècnic IT,0.0,False,NaN,NaN,15.1
996,996,Lucía,Roca Vila,21702378D,Espanya,Alacant,32.0,0.0,1940.0,NaN,1158€,NaN,True,Grup A,Data Analyst,NaN,NaN,NaN,NaN,19.4
897,897,Pol,Torres,25021199W,Espanya,Bilbao,28.0,12.0,1994.0,NaN,998 euros,True,NaN,Grup A,Responsable de vendes,1.0,NaN,NaN,NaN,16.3
300,300,Lucía,Domènech Álvarez,25145863X,Espanya,Madrid,14.0,4.0,1990.0,NaN,920 €,True,NaN,Grup A,Analista junior,3.0,True,80000.0,25.0,14.9
980,980,Marina,Romero Pujol,25303511D,Espanya,Palma,12.0,9.0,1997.0,NaN,"2,004 €",True,NaN,Grup B,Cap de projecte,2.0,False,NaN,NaN,15.9



Nombre de registres després d'eliminar els registres invàlids: 952

Files completament buides: 0
RESUM DE LA VALIDACIÓ
Nombre final de registres: 952
DNI únics: 952

Valors nuls restants:


,0
unnamed:_0,0
nom,0
cognoms,0
dni,0
pais_d'origen,0
ciutat,0
dia_de_naixement,0
mes_de_naixement,0
any_de_naixement,0
genere,0


## <font color = "pink"> Exercici 3: Transformacions necessàries i preparació del dataset </font>
🎯 Objectiu: preparar el dataset perquè sigui coherent, analitzable i estructuralment correcte abans de realitzar qualsevol anàlisi descriptiva.

En aquest exercici hauràs d'aplicar transformacions bàsiques però essencials, orientades a millorar la qualitat de les dades i facilitar l'anàlisi posterior.


Aspectes a revisar i transformar:


» Els valors categòrics estan normalitzats i són consistents?

» Els valors nuls són coherents amb la informació disponible?

» Hi ha redundància de variables que pugui generar problemes analítics?

» Les variables contextuals (com temperatura, salari) estan completes i preparades per a l'anàlisi?

Concretament, hauràs de:


» Normalitzar els valors categòrics per evitar ambigüitats (per exemple, valors que representen el mateix concepte però estan escrits de manera diferent).

» Revisar la coherència dels valors nuls, assegurant que:
Les persones que tenen quilometratge anual i consum mitjà associats disposen de cotxe. En els casos en què consti consum mitjà però no quilometratge anual, caldrà decidir com resoldre aquesta incoherència de manera justificada.

Les persones que tenen un valor en el nombre de fills efectivament tenen fills.

» Unificar la informació sobre fills en una única columna, ja que mantenir diverses columnes amb informació complementària no és òptim i pot generar problemes analítics (redundància o multicolinealitat).

» Validar la temperatura mitjana. En cas que aquesta informació falti, hauràs d'imputar-la utilitzant un criteri raonable (per ciutat o per un nivell superior justificat).

» Revisar i imputar els salaris faltants, aplicant criteris analítics coherents (com ara càrrec, grup professional, ciutat, gènere o franja d'edat).

 Important


en aquest exercici no es tracta d'eliminar registres, sinó de transformar, normalitzar i preparar les dades perquè siguin analitzables de manera rigorosa en els nivells següents.

In [7]:
# NORMALITZACIÓ DE VARIABLES CATEGÒRIQUES

columnes_text = df.select_dtypes(include="object").columns

for columna in columnes_text:
    df[columna] = (
        df[columna]
        .astype(str)
        .str.strip()
        .str.lower()
    )

# NORMALITZACIÓ DEL GÈNERE

df["genere"] = df["genere"].replace({

    "h": "home",
    "m": "home",
    "home": "home",

    "d": "dona",
    "f": "dona",
    "dona": "dona",

    "??": "desconegut",
    "nc": "desconegut",
    "a": "desconegut"

})

# NORMALITZACIÓ DE TE_COTXE

df["te_cotxe"] = (
    df["te_cotxe"]
    .replace({
        True: "si",
        False: "no",
        "true": "si",
        "false": "no",
        "sí": "si",
        "s": "si",
        "n": "no"
    })
)

print("Valors del gènere:")
display(df["genere"].value_counts())

# COHERÈNCIA DEL COTXE

# Si no té cotxe, no pot tenir km ni consum.

mask = df["te_cotxe"] == "no"

df.loc[mask, "km_anuals"] = pd.NA
df.loc[mask, "consum_mitja_l_100km"] = pd.NA

# Si hi ha consum però no quilometratge,
# imputem la mediana dels km.

mask = (
    df["consum_mitja_l_100km"].notna()
) & (
    df["km_anuals"].isna()
)

df.loc[mask, "km_anuals"] = df["km_anuals"].median()

# INFORMACIÓ DELS FILLS

df["nombre_fills"] = df["nombre_fills"].fillna(0)

df.loc[
    df["nombre_fills"] < 0,
    "nombre_fills"
] = 0

df["nombre_fills"] = df["nombre_fills"].astype(int)

# TEMPERATURA

df["temperatura_mitjana_ciutat"] = (
    df.groupby("ciutat")["temperatura_mitjana_ciutat"]
      .transform(lambda x: x.fillna(x.mean()))
)

df["temperatura_mitjana_ciutat"] = (
    df["temperatura_mitjana_ciutat"]
      .fillna(df["temperatura_mitjana_ciutat"].mean())
)

# CONVERTIR EL SALARI A NUMÈRIC

df["salari_mensual"] = (
    df["salari_mensual"]
    .astype(str)
    .str.replace("€", "", regex=False)
    .str.replace("€", "", regex=False)
    .str.replace("euros", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.strip()
)

df["salari_mensual"] = pd.to_numeric(
    df["salari_mensual"],
    errors="coerce"
)

# IMPUTACIÓ DELS SALARIS

# Primer pel càrrec

df["salari_mensual"] = (
    df["salari_mensual"]
    .fillna(
        df.groupby("ca_rrec")["salari_mensual"]
          .transform("median")
    )
)

# Després pel grup professional

df["salari_mensual"] = (
    df["salari_mensual"]
    .fillna(
        df.groupby("grup_professional")["salari_mensual"]
          .transform("median")
    )
)

# Finalment amb la mediana global

df["salari_mensual"] = (
    df["salari_mensual"]
    .fillna(
        df["salari_mensual"].median()
    )
)

# COMPROVACIÓ FINAL

print("=" * 50)
print("VALORS NULS RESTANTS")
display(df.isnull().sum())

print("\nTIPUS DE DADES")
display(df.dtypes)

print("\nRESUM DEL SALARI")
display(df["salari_mensual"].describe())

Valors del gènere:


,count
genere,
dona,543
home,361
desconegut,48


VALORS NULS RESTANTS


,0
unnamed:_0,0
nom,0
cognoms,0
dni,0
pais_d'origen,0
ciutat,0
dia_de_naixement,0
mes_de_naixement,0
any_de_naixement,0
genere,0



TIPUS DE DADES


,0
unnamed:_0,int64
nom,object
cognoms,object
dni,object
pais_d'origen,object
ciutat,object
dia_de_naixement,float64
mes_de_naixement,float64
any_de_naixement,float64
genere,object



RESUM DEL SALARI


,salari_mensual
count,952.000000
mean,1453.049720
std,693.914049
min,1.007000
25%,1028.500000
50%,1328.000000
75%,1804.750000
max,3672.000000


## <font color = "pink">Exercici 4: Primeres anàlisis descriptives </font>
🎯 Objectiu: obtenir una primera visió global del dataset un cop s'han aplicat les transformacions bàsiques, mitjançant estadístiques descriptives i comparacions simples entre grups rellevants.

Aspectes a revisar i analitzar:


» Construir taules resum descriptives.

» Fer comparacions bàsiques entre grups rellevants.

Concretament, hauràs de:


» Construir taules resum descriptives de salaris com una taula resum que permeti veure el sou mig, medià, mínim i màxim per Gènere. Ordena la taula en funció del sou mig.

» Crear una taula resum amb el salari mig per gènere (files) i país d'origen (columnes). Afegeix-hi les mitjanes als marges de la taula. Aplica format condicional a la taula per veure en un color més intens els valors més elevats.

In [8]:
# TAULA RESUM DELS SALARIS PER GÈNERE

resum_salaris = (
    df.groupby("genere")["salari_mensual"]
      .agg(
          Salari_Mitja="mean",
          Mediana="median",
          Minim="min",
          Maxim="max"
      )
      .sort_values("Salari_Mitja", ascending=False)
      .round(2)
)

print("Resum de salaris per gènere:")

display(resum_salaris)

# TAULA DINÀMICA

taula = pd.pivot_table(
    df,
    values="salari_mensual",
    index="genere",
    columns="pais_d'origen",
    aggfunc="mean",
    margins=True,
    margins_name="Mitjana"
)

print("\nSalari mitjà per gènere i país d'origen:")

display(taula.round(2))

# FORMAT CONDICIONAL

display(
    taula.style
         .background_gradient(cmap="Purples")
         .format("{:.2f}")
)

# EXTRA - PERCENTATGE DE PERSONES PER GÈNERE

percentatges = (
    df["genere"]
      .value_counts(normalize=True)
      .mul(100)
      .round(2)
      .rename("Percentatge")
)

print("\nPercentatge de persones per gènere:")

display(percentatges)

# EXTRA - RÀTIO SALARIAL

mitjana_general = df["salari_mensual"].mean()

ratio = (
    df.groupby("genere")["salari_mensual"]
      .mean()
      .div(mitjana_general)
      .round(2)
      .rename("Ràtio")
)

print("\nRàtio salarial respecte la mitjana general:")

display(ratio)

Resum de salaris per gènere:


,Salari_Mitja,Mediana,Minim,Maxim
genere,,,,
home,1581.44,1429.0,1.01,3672.0
desconegut,1453.15,1267.0,1.10,3600.0
dona,1367.68,1309.0,1.01,3564.0



Salari mitjà per gènere i país d'origen:


pais_d'origen,alemanya,espanya,franã§a,itã lia,narnia,noruega,Mitjana
genere,,,,,,,
desconegut,1634.00,1439.77,1266.42,2047.40,1328.0,1.24,1453.15
dona,1326.16,1356.95,1353.61,1384.83,NaN,1578.95,1367.68
home,1726.95,1598.52,1520.64,1530.39,NaN,1451.78,1581.44
Mitjana,1430.98,1455.31,1409.02,1470.11,1328.0,1489.02,1453.05


pais_d'origen,alemanya,espanya,franã§a,itã lia,narnia,noruega,Mitjana
genere,,,,,,,
desconegut,1634.00,1439.77,1266.42,2047.40,1328.00,1.24,1453.15
dona,1326.16,1356.95,1353.61,1384.83,nan,1578.95,1367.68
home,1726.95,1598.52,1520.64,1530.39,nan,1451.78,1581.44
Mitjana,1430.98,1455.31,1409.02,1470.11,1328.00,1489.02,1453.05



Percentatge de persones per gènere:


,Percentatge
genere,
dona,57.04
home,37.92
desconegut,5.04



Ràtio salarial respecte la mitjana general:


,Ràtio
genere,
desconegut,1.00
dona,0.94
home,1.09
